File: monte_carlo.jl
function findPi(num_samples, trials_per_sample)
  samples = Array{Float64}(undef, num_samples)
  mean = 0.0

  for sample = 1:num_samples
    samples[sample] = integrateCircle(trials_per_sample)
  end

  mean = sum(samples) / size(samples)[1]

  std_dev = getStdDev(mean, samples)

  return mean, std_dev
end

function getStdDev(mean, samples)
  variance = 0

  for sample in samples
    variance += (sample - mean) ^ 2
  end

  variance /= size(samples)[1]
  return sqrt(variance)
end

function integrateCircle(points_placed)
  counter = 0

  for trial = 1:points_placed
    y_pos = rand()
    x_pos = rand()

    dist = y_pos ^ 2 + x_pos ^ 2
    if dist <= 1
      counter += 1
    end
  end

  return 4.0 * counter / points_placed
end

function monte_carlo()
  res = [0.0, 0.0]
  for i in 1:20
    res1, res2 = findPi(1000, 50000)
    res[1] += res1 / 20.0
    res[2] += res2 / 20.0
  end
  print(res)
end

File: monte_carlo_go/monte_carlo.go
package main

import (
	"fmt"
	"math"
	"math/rand"
)

func getStdDev(mean float64, samples []float64) float64 {
	numSamples := len(samples)
	variance := 0.0

	for i := 0; i < numSamples; i++ {
		variance += (samples[i] - mean) * (samples[i] - mean)
	}
	return math.Sqrt(variance / float64(numSamples))
}

func integrateCircle(pointsPlaced int) float64 {
	counter := 0

	for i := 0; i < pointsPlaced; i++ {
		yPos := rand.Float64()
		xPos := rand.Float64()

		if yPos*yPos+xPos*xPos <= 1.0 {
			counter++
		}
	}

	return 4.0 * float64(counter) / float64(pointsPlaced)
}

func findPi(numSamples int, trialsPerSample int) (float64, float64) {
	samples := make([]float64, numSamples)
	mean := 0.0

	for sample := 0; sample < numSamples; sample++ {
		samples[sample] = integrateCircle(trialsPerSample)
		mean += samples[sample] / float64(numSamples)
	}

	stdDev := getStdDev(mean, samples)

	return mean, stdDev
}

func main() {
	resMean := 0.0
	resStdDev := 0.0
	for i := 0; i < 20; i++ {
		mean, stdDev := findPi(1000, 50000)
		resMean += mean
		resStdDev += stdDev
	}
	resMean /= 20.0
	resStdDev /= 20.0
	fmt.Println(resMean, resStdDev)
}
